src/frontend/src/binder/mod.rs

```diff
@@ -420,27 +420,21 @@ impl Binder {
         self.included_relations.clone()
     }
 
-    fn push_context(&mut self) {
+    #[must_use]
+    fn push_context(&mut self) -> PopContextGuard<'_> {
         let new_context = std::mem::take(&mut self.context);
         self.context
             .cte_to_relation
             .clone_from(&new_context.cte_to_relation);
         let new_lateral_contexts = std::mem::take(&mut self.lateral_contexts);
         self.upper_subquery_contexts
             .push((new_context, new_lateral_contexts));
-    }
 
-    fn pop_context(&mut self) -> Result<()> {
-        let (old_context, old_lateral_contexts) = self
-            .upper_subquery_contexts
-            .pop()
-            .ok_or_else(|| ErrorCode::InternalError("Popping non-existent context".to_string()))?;
-        self.context = old_context;
-        self.lateral_contexts = old_lateral_contexts;
-        Ok(())
+        PopContextGuard(self)
```

Comment:
nit: add an explicit constructor for the context guard may be better for readability.

Reply:
It's not a good practice, there should be the only path to construct a Guard.